### ISSUE TITLE
[BUG] CSV parser copies the host buffer on the host

### ISSUE THREAD

**Issue by jlowe** [rapidsai/cudf#2800]
When passing a host buffer to `read_csv`, the code makes a couple of host-side copies of the buffer.  The buffer is often quite large which can lead to hundreds of milliseconds spent on the copies per invocation (plus the additional memory footprint).

The first copy occurs in this code in `read_csv`:
```c++
    if (options.input_data_form == HOST_BUFFER) {
      options.filepath_or_buffer =
          std::string(args.source.buffer.first, args.source.buffer.second);
```

The second copy occurs when `csv::reader::Impl` is constructed from the options object constructed above.  The `args_` member gets copy-constructed and the `filepath_or_buffer` field copies it again.

**Comment by j-ieong** [comment]
I'll fix this in the same PR as #2493, where we refactor the CSV to align with the behavior of Parquet and ORC readers.

**Comment by jlowe** [comment]
Fixed by #2838
